app/scripts/modules/core/securityGroup/securityGroupWriter.service.ts

```diff
@@ -0,0 +1,70 @@
+import {assignWith, isUndefined} from 'lodash';
+import {module} from 'angular';
+
+import {TASK_EXECUTOR, TaskExecutor, IJob} from 'core/task/taskExecutor';
+import {INFRASTRUCTURE_CACHE_SERVICE, InfrastructureCacheService} from 'core/cache/infrastructureCaches.service';
+import {ISecurityGroup} from 'core/domain';
+import {Application} from 'core/application/application.model';
+import {ITask} from 'core/task/task.read.service';
+
+export interface ISecurityGroupJob extends IJob {
+  credentials?: string;
+  regions?: string[];
+  securityGroupName?: string;
+}
+export class SecurityGroupWriterService {
+
+  static get $inject(): string[] {
+    return ['infrastructureCaches', 'taskExecutor'];
+  }
+
+  constructor(private caches: InfrastructureCacheService,
+              private executor: TaskExecutor) {}
+
+  public deleteSecurityGroup(securityGroup: ISecurityGroup,
+                             application: Application,
+                             params: ISecurityGroupJob): ng.IPromise<ITask> {
+
+    params.type = 'deleteSecurityGroup';
+    params.securityGroupName = securityGroup.name;
+    params.regions = [securityGroup.region];
+    params.credentials = securityGroup.accountId;
+
+    const operation: ng.IPromise<ITask> = this.executor.executeTask({
+      job: [params],
+      application: application,
+      description: `Delete Security Group: ${securityGroup.name}`
+    });
+    this.caches.clearCache('securityGroups');
+
+    return operation;
+  }
+
+  public upsertSecurityGroup(securityGroup: ISecurityGroup,
+                             application: Application,
+                             description: string,
+                             params: ISecurityGroupJob): ng.IPromise<ITask> {
+
+    params.type = 'upsertSecurityGroup';
+    params.credentials = securityGroup.credentials || securityGroup.accountName;
+
+    // We want to extend params with all attributes from command, but only if they don't already exist.
+    assignWith(params, securityGroup, (value: any, other: any) => {
+      return isUndefined(value) ? other : value;
+    });
```

Comment:
So this one is interesting because `Object.assign` will _overwrite_ previously populated values if a property in a _later_ object is explicitly set to `undefined`.  I.e.,
```javascript
var foo = { bar: 'bat', baz: 'quack' };
var bar = { baz: 'snoopy', bar: undefined };
Object.assign(foo, bar); // { bar: undefined, baz: "snoopy" }
```
Now, normally we don't set property values explicitly to `undefined` - if we need to delete them we use `delete`.  But, the fact that we're not only using `assignWith`, but also explicitly checking for `undefined` to ensure the target object does not get properties overwritten, gives me pause about removing this.  Thoughts?